### ISSUE TITLE
When converting class to callable, pyright doesn't honor non-standard `__new__` return type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#7686]
The newly-updated typing spec indicates that when converting a class to a callable, a type checker should determine whether the `__new__` method returns a type other than the class (or subclass) that is being constructed. If not, it should ignore the `__init__` method if present. Pyright doesn't do this currently.

Here's the test case that currently fails.

```python
class Class6Proxy: ...


class Class6:
    """__new__ that causes __init__ to be ignored"""

    def __new__(cls) -> Class6Proxy:
        return Class6Proxy.__new__(cls)

    def __init__(self, x: int) -> None:
        """This __init__ is ignored for purposes of conversion"""
        pass


class Expected6(Protocol):
    def __call__(self) -> Class6Proxy: ...


r6 = accepts_callable(Class6)
reveal_type(r6)  # ``def () -> Class6Proxy``
t6: Expected6 = r6  # OK
assert_type(r6(), Class6Proxy)
r6(1)  # E
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
